rholang/src/main/scala/coop/rchain/rholang/interpreter/Splittable.scala

```diff
@@ -0,0 +1,25 @@
+package coop.rchain.rholang.interpreter
+
+/**
+  * This is a lawless typeclass to abstract over types providing a `takeFirst()` method, as defined in
+  * NonDetFreeMapWithCost's ops.
+  *
+  * @tparam F
+  */
+trait Splittable[F[_]] {
```

Comment:
Adding this in complete separation without tests or usage makes it hard to capture the intent. This commit should be squashed with those containing usage and tests

Reply:
There aren't tests. The usage is already there in SpatialMatcher - yet not via a typeclass (`allMatches.takeFirst()`, `allMathches` is of type `NonDetFreeMatchWithCost`). After changing from `type F = NonDetFreeMatchWithCost` to `F: Splittable` the code still compiles.

Splittable actually looks similar to a dual of MonoidK - so ComonoidK? It's not in cats though and when I asked, Luka Jacobovitz said it'd probably hard to have proper laws for that.

My proposal is we leave this as is. The only place I could logically squash that too is the already-huge last commit. WDYT?

Reply:
Fair enough. It should definitely be squashed together with a589606b96153484616762c27903b11f082b0a42. Let's leave the "already huge commit" as is

Reply:
`ComonoidK` would have two functions `cocombine[A](fa: F[A]): (F[A],F[A])` and `coempty[A](fa: F[A]): Unit`. If it's really desired, I don't see why it would be any harder to have laws than for `MonoidK`.

Reply:
But that would really only be helpful if you want to return the head of `NonDetFreeMatchWithCost` along with the dropped tail.

Reply:
Ignoring the tail is not an issue. I'm though not sure if we can get away with having `cocombine` (or `split`?) return `(h, t)` for a list of `h :: t`). My hunch is that the split is a bit arbitrary and it is unclear what (if any) laws we could have for interactions with the Monoid part.

Anyway, I'm sure both our codebase, and the cats project, would enjoy a `ComonoidK` typeclass with a nice set of laws and at least a few imaginable useful instances :) I'll be happy to review such a PR and/or help make it happen if we can first sort out the laws :)

Reply:
FWIW here's the discussion in cats gitter, with each reply being less comprehensible for me than the previous one :sweat_smile: Maybe it'll spark some ideas :) https://gitter.im/typelevel/cats?at=5c37d0609289cc7aa753882d